### PR TITLE
🏗 add storybook for react

### PR DESCRIPTION
**Pull request by dethstrobe**
In order to test the react components; this PR adds a new storybook that uses the react transpiled components.

- added `amp storybook-react` command
- cleaned up some code for amp-dailymotion

### CODE REVIEW

File: build-system/tasks/storybook-react/OWNERS
Comment (rcebulko) [inline build-system/tasks/storybook-react/OWNERS:17]: Should this apply to the whole directory, rather than the list of groups above (given they overlap significantly)?

File: build-system/tasks/storybook-react/README.md
Comment (rcebulko) [inline build-system/tasks/storybook-react/README.md:52]: Question for myself, which may lead to adding docs. Would this have the same effect?
```jsx
MyStory.args = {
  width: 500,
  height: 200,
  layout: 'responsive',
  data: { foo: 'bar' },
};
```

File: build-system/tasks/storybook-react/README.md
Comment (rcebulko) [inline build-system/tasks/storybook-react/README.md:153]: Super-nit/trivia: this is `PascalCase` with the leading capitalization :)

File: build-system/tasks/storybook-react/index.js
Comment (rcebulko) [inline build-system/tasks/storybook-react/index.js:22]: Suggestion since we're branching 3x on `build`, feel free to ignore if you find it less clear

```suggestion
  if (build) {
    envs.forEach(buildEnv);
    return;
  }
  
  if (envs.includes('amp')) {
    await runAmpDevBuildServer();
  }
  createCtrlcHandler('storybook');
  envs.forEach(launchEnv);
 
```

File: build-system/tasks/storybook-react/index.js
Comment (rcebulko) [inline build-system/tasks/storybook-react/index.js:34]: Consider adding an example of allowed values (namely `amp` and `preact`, along with an explanation that the expected format is comma-delimited. Appending (`default: "amp,preact"`) might suffice

File: build-system/tasks/storybook-react/package.json
Comment (rcebulko) [inline build-system/tasks/storybook-react/package.json:4]:
```suggestion
  "description": "amp-storybook tasks",
```

File: build-system/tasks/storybook-react/package.json
Comment (rcebulko) [inline build-system/tasks/storybook-react/package.json:5]: `main` is `index.js` by default
```suggestion
```

File: build-system/tasks/storybook-react/react-env/main.js
Comment (rcebulko) [inline build-system/tasks/storybook-react/react-env/main.js:1]: What forbidden terms are used in this file? Prefer `eslint-disable-next-line` or adding this file to the forbidden-terms allowlist for the rule it's violating (usually preferred)

File: build-system/tasks/storybook-react/react-env/main.js
Comment (rcebulko) [inline build-system/tasks/storybook-react/react-env/main.js:10]: nit: not async
```suggestion
  webpackFinal: (config) => {
```

File: build-system/tasks/storybook-react/react-env/main.js
Comment (rcebulko) [inline build-system/tasks/storybook-react/react-env/main.js:9]:
```suggestion
  managerWebpack: webpackConfigNoChunkTilde,
```

File: build-system/tasks/storybook-react/react-env/webpack.config.js
Comment (rcebulko) [inline build-system/tasks/storybook-react/react-env/webpack.config.js:24]: Two notes:
1. Why are these aliases needed here?
2. Just confirming, have you tested using these aliases (both compiling/running and within VSCode/linting)? Aliases into `dist` sometimes have issues with the linter since the files themselves often don't exist

File: build-system/tasks/storybook-react/react-env/webpack.config.js
Comment (rcebulko) [inline build-system/tasks/storybook-react/react-env/webpack.config.js:60]: It's possible the linter won't allow this, but the amount of vertical space demands I at least try 😛 
```suggestion
            ['babel/preset-react', {'runtime': 'automatic'}],
```

File: build-system/tasks/storybook/index.js
Comment (rcebulko) [inline build-system/tasks/storybook/index.js:25]: Consider doing something like the following:
```javascript
/** @typedef {'amp'|'preact'|'react'} StorybookEnv */
```
or declaring an `_Enum` type

File: extensions/amp-base-carousel/1.0/test/test-amp-base-carousel.js
Comment (rcebulko) [inline extensions/amp-base-carousel/1.0/test/test-amp-base-carousel.js:389]:
```suggestion
    it('should start at slide 1 with slide attr set to 1', async () => {
```

File: extensions/amp-base-carousel/1.0/test/test-amp-base-carousel.js
Comment (rcebulko) [inline extensions/amp-base-carousel/1.0/test/test-amp-base-carousel.js:52]: nit: simplicity
```suggestion
        .map((wrapper) => wrapper.querySelector('slot'))
```

File: build-system/tasks/storybook-react/README.md
Comment (kvchari) [inline build-system/tasks/storybook-react/README.md:45]: should this just be `my-component`?

File: build-system/tasks/storybook-react/react-env/webpack.config.js
Comment (rcebulko) [inline build-system/tasks/storybook-react/react-env/webpack.config.js:29]: @alanorozco WDYT about adding `#extensions` as a root alias at this point?

File: build-system/tasks/storybook-react/README.md
Comment (kvchari) [inline build-system/tasks/storybook-react/README.md:205]: change amp css selector to something that make sense for preact (or remove)

File: extensions/amp-lightbox-gallery/1.0/storybook/Basic.js
Comment (rcebulko) [inline extensions/amp-lightbox-gallery/1.0/storybook/Basic.js:14]: These comments may be lost/shifted when eslint auto-sorts+groups imports

File: build-system/tasks/storybook-react/README.md
Comment (kvchari) [inline build-system/tasks/storybook-react/README.md:214]: change `amp-` to something that make sense for preact (or remove)

File: extensions/amp-sidebar/1.0/storybook/Basic.js
Comment (rcebulko) [inline extensions/amp-sidebar/1.0/storybook/Basic.js:8]: I wonder if we can't do this automatically during compilation somehow? I'm sure Alan will weigh in when he reviews

File: build-system/tasks/storybook-react/index.js
Comment (kvchari) [inline build-system/tasks/storybook-react/index.js:12]: Did you mean this:
```suggestion
  // React dist needs to be compiled before running
```

File: build-system/tasks/storybook-react/index.js
Comment (kvchari) [inline build-system/tasks/storybook-react/index.js:14]: Should this be `react`?
```suggestion
  const {build = false, 'storybook_env': env = 'amp,react'} = argv;
```

File: build-system/tasks/storybook-react/index.js
Comment (kvchari) [inline build-system/tasks/storybook-react/index.js:20]: I think this needs to be the name of the command:
```suggestion
    createCtrlcHandler('storybook-react');
```

File: build-system/tasks/storybook-react/react-env/webpack.config.js
Comment (kvchari) [inline build-system/tasks/storybook-react/react-env/webpack.config.js:22]: Do we need to alias every extension's binaries manually? If so, not a huge fan since we're basically aliasing `**/extensions/*/1.0/component` -> `**/extensions/*/1.0/dist/component-react.max`. Is there a simpler way of doing this? Would it be simple to use a custom webpack-plugin? (Maybe this logic could be reused for running unit tests on react?)

File: extensions/amp-render/1.0/storybook/Basic.amp.js
Comment (kvchari) [inline extensions/amp-render/1.0/storybook/Basic.amp.js:1]: Why did we need to delete the the amp-render storybooks?

File: extensions/amp-base-carousel/1.0/test/test-amp-base-carousel.js
Comment (kvchari) [inline extensions/amp-base-carousel/1.0/test/test-amp-base-carousel.js:389]: did you want to remove the `.only`?

File: build-system/tasks/storybook/index.js
Comment (kvchari) [inline build-system/tasks/storybook/index.js:19]: I vaguely remember we discussing this in person - was it possible to automatically spin up the react storybook along with the preact storybook in the `amp storybook` command?

**Review by kvchari — COMMENTED** [review]
Looks good - had a couple questions for more info and doc fixes

File: build-system/tasks/storybook-react/react-env/webpack.config.js
Comment (rcebulko) [reply build-system/tasks/storybook-react/react-env/webpack.config.js:22]: Weighing in a little: Aliases are tricky to get overly complicated with because many different systems (linter, VSCode auto-import/reference jumping, compiler, TypeScript, and webpack) all need to be on the same page, and some implement differently than others. Likely possible to do, but should probably be tackled as an isolated change given that it's no unlikely to introduce regressions in one or more of the areas and will be easier/safer to revert if that happens as an isolated change to import resolution ratherhan this whole PR

File: build-system/tasks/storybook-react/index.js
Comment (dethstrobe) [reply build-system/tasks/storybook-react/index.js:22]: I actually like this a lot more

File: build-system/tasks/storybook-react/OWNERS
Comment (dethstrobe) [reply build-system/tasks/storybook-react/OWNERS:17]: would it just look like this?

```suggestion
      pattern: '**/*',
```

File: build-system/tasks/storybook-react/README.md
Comment (dethstrobe) [reply build-system/tasks/storybook-react/README.md:52]: I feel like we alias this amp attributes to preact props, so I don't think it'd map that smoothly. But I'm not clear on how all this all works under the hood.

File: build-system/tasks/storybook-react/index.js
Comment (dethstrobe) [reply build-system/tasks/storybook-react/index.js:22]: I like this a lot more. It's a good refactor.

File: amp.js
Comment (rsimha) [inline amp.js:64]: The `amp storybook` task has a flag called `--storybook_env` which takes `amp` or `preact` as modes. 

image

Curious about why we're adding a new task here instead of introducing a new mode to the existing flag. Are the code paths substantially different? If so, it may be possible to implement the guts in a separate file while retaining the original task name.

File: build-system/tasks/storybook-react/package.json
Comment (rsimha) [inline build-system/tasks/storybook-react/package.json:2]: This is sharing a name with build-system/tasks/storybook/package.json. Maybe change it to include "react"?

(Unless you agree with my first comment, in which case this file can be removed.)

File: build-system/tasks/storybook-react/package.json
Comment (dethstrobe) [reply build-system/tasks/storybook-react/package.json:4]: I'm going to put this as 
```suggestion
  "description": "storybook-react tasks to start bento components using their react binaries",
```
thinking it over it's not really for amp. It's more for bento.

**Review by rsimha — COMMENTED** [review]
A couple high-level comments. Will defer to Ryan, Keshav, et. al. for a more detailed review.

File: build-system/tasks/storybook-react/react-env/main.js
Comment (dethstrobe) [reply build-system/tasks/storybook-react/react-env/main.js:1]: I forgot why I added that there, but it doesn't appear to be needed anymore.

Comment (dethstrobe) [reply build-system/tasks/storybook-react/react-env/main.js:1]: If i recall, there was a linting rule that prevented me from using the `@storybook/addon-knobs` but it looks like I can now.

File: build-system/tasks/storybook-react/OWNERS
Comment (rcebulko) [reply build-system/tasks/storybook-react/OWNERS:17]: It would be the rule above, with no `pattern` specified at all

File: build-system/tasks/storybook-react/index.js
Comment (dethstrobe) [reply build-system/tasks/storybook-react/index.js:34]: This command is pretty much only going to have react for right now.

File: build-system/tasks/storybook-react/react-env/webpack.config.js
Comment (dethstrobe) [reply build-system/tasks/storybook-react/react-env/webpack.config.js:24]: Yeah, I have the binaries build before running react storybook. So there should always be a dist directory.

The `amp-accordion` itself is used in a few storybooks, which I don't think it required, but might have been done to make the storybook examples look easier to understand.

I thought about just removing the use of accordion from those storybook examples, but I didn't feel safe making that refactor at the time and thought this aliasing would be a quicker and safer solution.

File: build-system/tasks/storybook-react/react-env/webpack.config.js
Comment (dethstrobe) [reply build-system/tasks/storybook-react/react-env/webpack.config.js:60]: It appears to be working.

File: extensions/amp-lightbox-gallery/1.0/storybook/Basic.js
Comment (dethstrobe) [reply extensions/amp-lightbox-gallery/1.0/storybook/Basic.js:14]: Hm...should i just remove the comments exampling them?

File: extensions/amp-sidebar/1.0/storybook/Basic.js
Comment (dethstrobe) [reply extensions/amp-sidebar/1.0/storybook/Basic.js:8]: I couldn't figure it out, but if we can, that's be wonderful.

File: amp.js
Comment (dethstrobe) [reply amp.js:64]: storybook adds all `@storybook` packages, which creates conflicts between the amp, preact, and react versions of storybook. This was the only way I could figure out how to make them play nice with each other.

**Review by rsimha — COMMENTED** [review]
As mentioned offline, we run `amp storybook --build` so that its output files are included in the output uploaded to the PR deploy bot. Might be worth including react mode too.

https://github.com/ampproject/amphtml/blob/5494590583195d57a6fb9d2ce6e3bd5d36828df2/build-system/pr-check/pr-deploy.js#L31

File: amp.js
Comment (rcebulko) [reply amp.js:64]: Workaround, curious what y'all's thoughts are: What if we provide a `amp storybook-react-internal` task and then extend the primary `amp storybook` task to defer to it when `--storybook_env` includes `react`, this way we can centralize some of the logic/have a single standard entrypoint, even if technical challenges force it to be implemented separately?

File: build-system/tasks/storybook-react/README.md
Comment (dethstrobe) [reply build-system/tasks/storybook-react/README.md:198]: It appears they're all Pascal cased in the `extensions/**/storybook/` directories.

File: build-system/tasks/storybook-react/index.js
Comment (dethstrobe) [reply build-system/tasks/storybook-react/index.js:14]: It should only be react. Oddly, I thought I did change this.

File: build-system/tasks/storybook-react/react-env/webpack.config.js
Comment (dethstrobe) [reply build-system/tasks/storybook-react/react-env/webpack.config.js:22]: I agree that all these aliases are pretty gross, but I couldn't find a better solution in a timely manner. We can try to investigate again later for something more elegant.

File: extensions/amp-render/1.0/storybook/Basic.amp.js
Comment (dethstrobe) [reply extensions/amp-render/1.0/storybook/Basic.amp.js:1]: It is not published to npm, so never have a react binary built. Which means this storybook story will fail to build.

File: build-system/tasks/storybook/index.js
Comment (dethstrobe) [reply build-system/tasks/storybook/index.js:19]: I was not able to get that working. For right now you need to use `amp storybook-react` to start it up

File: extensions/amp-render/1.0/storybook/Basic.amp.js
Comment (kvchari) [reply extensions/amp-render/1.0/storybook/Basic.amp.js:1]: Ah gotcha. Shouldn't we keep the amp storybook though? Or does the amp storybooks now use react under-the-hood?

File: build-system/tasks/storybook-react/react-env/webpack.config.js
Comment (rcebulko) [reply build-system/tasks/storybook-react/react-env/webpack.config.js:24]: > Yeah, I have the binaries build before running react storybook. So there should always be a dist directory.

Exactly, you have them built. If they aren't built already though, it may cause problems in linting/editor. Does this pass `amp lint`? This also masks that the files are dependent on build files (which the `/dist` path previously made visible). It's not clear to me that adding these aliases offers much benefit

File: build-system/tasks/storybook-react/react-env/webpack.config.js
Comment (rcebulko) [reply build-system/tasks/storybook-react/react-env/webpack.config.js:22]: To clarify, a better solution than what? Are the aliases needed to make it work for some reason?

File: amp.js
Comment (alanorozco) [reply amp.js:64]: @rcebulko We can have a similar scheme without the necessity of subtasks.

The issue is a package install conflict in a case where Storybook does introspection. The workaround needs a separation of `package.json` files, a simpler consolidation proposal would look like:

```diff
- build-system/tasks/storybook-react/
+ build-system/tasks/storybook/react-env/
- build-system/tasks/storybook/package.json
+ build-system/tasks/storybook/amp-env/package.json 
+ build-system/tasks/storybook/preact-env/package.json 
```

It's probably okay if we consolidate after merging this PR, though.

File: extensions/amp-accordion/1.0/storybook/Basic.js
Comment (alanorozco) [inline extensions/amp-accordion/1.0/storybook/Basic.js:10]: We'll need to add this to the extension generator template. 
On `build-system/tasks/make-extension/template/bento/extensions/amp-__component_name_hyphenated__/__component_version__/storybook/Basic.js`:

```diff
  import * as Preact from '#preact';
  import {Bento__component_name_pascalcase__} from '../component'
+ 
+ __jss_import_storybook__
```

On `build-system/tasks/make-extension/index.js`:

```diff
    ...(!options.nojss
      ? {
          '__jss_import_component_css__': `import {CSS as COMPONENT_CSS} from './component.jss'`,
          '__jss_component_css__': 'COMPONENT_CSS',
          '__jss_import_use_styles__': `import {useStyles} from './component.jss'`,
          '__jss_styles_use_styles__': 'const styles = useStyles()',
          '__jss_styles_example_or_placeholder__':
            '`${styles.exampleContentHidden}`',
+         '__jss_import_storybook__': `import '../component.jss'`,
        }
      : {
          '__jss_import_component_css_': '',
          '__jss_component_css__': 'null',
          '__jss_import_use_styles__': '',
          '__jss_styles_use_styles__': '',
          '__jss_styles_example_or_placeholder__': `'my-classname'`,
+         '__jss_import_storybook__': '',
        }),
```

File: extensions/amp-sidebar/1.0/storybook/Basic.js
Comment (alanorozco) [reply extensions/amp-sidebar/1.0/storybook/Basic.js:7]: It probably should be included explicitly since some files require imports from unexpected paths, like `import '../../../amp-base-carousel/1.0/component.jss';`.

Adding to the extension template as I mentioned above is somewhat equivalent for this purpose.

File: build-system/tasks/storybook-react/react-env/webpack.config.js
Comment (dethstrobe) [reply build-system/tasks/storybook-react/react-env/webpack.config.js:22]: Yes, because the stories for storybook are importing the preact components directly, we need to map them to the react versions or else the react storybook won't load.

File: build-system/tasks/storybook/index.js
Comment (rcebulko) [inline build-system/tasks/storybook/index.js:81]: Looking at this, I had a hard time following the nested backticks and `${...}`. Suggestion for readability/making it less likely someone like me breaks this somehow:

```suggestion
      const parameters = {
         ampBaseUrlOptions: [`${getBaseUrl()}/dist`],
       };
      const previewFileContents = [
          // eslint-disable-next-line local/no-forbidden-terms
          '// DO NOT SUBMIT',
          '// This preview.js file was generated for a specific PR build.',
          // JSON.stringify here. prevents XSS and other types of garbling.
          `export const parameters = (${JSON.stringify(parameters)});`,
        ].join('\n');
      writeFileSync(`${envConfigDir(env)}/preview.js`, previewFileContents);
```

It's a bit bulkier (esp. splitting params/contents into separate vars) but might be more resilient to future edits, especially to `parameters`

Also, it avoids extra leading whitespace on lines 2 and 3 in the preview. Which doesn't matter AT ALL but is still slightly satisfying 🙃

File: build-system/tasks/storybook-react/react-env/webpack.config.js
Comment (rcebulko) [reply build-system/tasks/storybook-react/react-env/webpack.config.js:22]: Ahh gotcha. This seems like it could become increasingly unwieldy as the number of Bento components grows, but I don't see any clear alternatives without diving deep. Sounds like a job for another PR, thanks for entertaining my inquisition!

File: build-system/tasks/storybook-react/index.js
Comment (dethstrobe) [reply build-system/tasks/storybook-react/index.js:47]: The Error class seems to only take one argument, which is a string.

Comment (rcebulko) [reply build-system/tasks/storybook-react/index.js:48]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Error#:~:text=When%20catching%20and%20re%2Dthrowing%20an%20error%20with%20a%20more%2Dspecific%20or%20useful%20error%20message%2C%20this%20property%20should%20be%20used%20to%20pass%20the%20original%20error.

It has a few signatures. Looks like I was also wrong; should be
```
```suggestion
  ).on('error', (err) => {
    throw new Error('Launch failed', {cause: err});
  });
```

Comment (dethstrobe) [reply build-system/tasks/storybook-react/index.js:48]: I'm getting is typescript error
```
Expected 0-1 arguments, but got 2.ts(2554)
```
but i'm just going to ignore it since our linter doesn't seem to mind.

Comment (rcebulko) [reply build-system/tasks/storybook-react/index.js]: Nice find! It appears MS Edge doesn't support this param:
image

Maybe the best we can do here then is
`throw new Error(`Launch failed: ${err}`);`

File: extensions/amp-render/1.0/storybook/Basic.amp.js
Comment (dethstrobe) [reply extensions/amp-render/1.0/storybook/Basic.amp.js:1]: amp storybook doesn't actually work.

File: build-system/tasks/make-extension/template/bento/extensions/amp-__component_name_hyphenated__/__component_version__/storybook/Basic.js
Comment (alanorozco) [inline build-system/tasks/make-extension/template/bento/extensions/amp-__component_name_hyphenated__/__component_version__/storybook/Basic.js:4]: nit: 
```suggestion
__jss_import_storybook__

```

File: build-system/tasks/storybook-react/README.md
Comment (alanorozco) [reply build-system/tasks/storybook-react/README.md:52]: @dethstrobe That's correct

File: build-system/tasks/storybook-react/index.js
Comment (alanorozco) [reply build-system/tasks/storybook-react/index.js:70]: Good call. This file is going away after moving `package.json` files, so we may address this separately as a change to `build-system/tasks/storybook/index.js` instead.

File: build-system/tasks/storybook-react/index.js
Comment (alanorozco) [reply build-system/tasks/storybook-react/index.js:103]: Yep, this file is going away also.

File: build-system/tasks/storybook-react/react-env/webpack.config.js
Comment (alanorozco) [reply build-system/tasks/storybook-react/react-env/webpack.config.js:29]: These paths are ugly for sure. But I'm not sure that making them prettier with an alias is the answer. 

The specific items in dist list can go away because there's a pattern. A [`ResolverPlugin`](https://github.com/webpack/docs/wiki/list-of-plugins#resolverplugin) can be configured to follow it:

```
${x}/component -> ${x}/dist/component-react.max
${x}/component.jss -> ${x}/dist/styles.css
```

**Review by alanorozco — APPROVED** [review]
LGTM with a couple of tasks to do after: task consolidation and `ResolverPlugin` to remove hardcoded extension aliases. 

Before merging, please:
1. undo changes to `build-system/tasks/storybook/index.js`
2. Add TODO comments for tasks mentioned.

File: build-system/tasks/storybook-react/index.js
Comment (alanorozco) [inline build-system/tasks/storybook-react/index.js:47]: Looks like typecheck doesn't like `{cause: err}`. I'd say it's ok to remove `cause`.

https://app.circleci.com/pipelines/github/ampproject/amphtml/21594/workflows/f5a66572-3edb-4cd2-a7bb-5f18c4054ed3/jobs/421912?invite=true#step-116-211

**Review by rsimha — APPROVED** [review]
Owners approval for infra.

**Pull request merged**
